### ISSUE TITLE
Postgresql10 database shutdowns after  CRUD operation is performed when pgaudit is enabled

### ISSUE THREAD

**Issue by ramakanthreddy** [pgaudit/pgaudit#92]
Below is the log captured from pg log file.

When select/insert operations are performed the database goes down. but the pg-audit logs are present in log file. Database works fine when pgaudit is disabled.

< 2020-04-27 09:06:04.076 UTC    13967 5ea6a061.368f 2020-04-27 09:05:37 UTC  >LOG:  received fast shutdown request
< 2020-04-27 09:06:04.080 UTC    13967 5ea6a061.368f 2020-04-27 09:05:37 UTC  >LOG:  aborting any active transactions
< 2020-04-27 09:06:04.083 UTC    13967 5ea6a061.368f 2020-04-27 09:05:37 UTC  >LOG:  worker process: logical replication launcher (PID 13975) exited with exit code 1
< 2020-04-27 09:06:04.083 UTC    13970 5ea6a061.3692 2020-04-27 09:05:37 UTC  >LOG:  shutting down
< 2020-04-27 09:06:04.084 UTC    13970 5ea6a061.3692 2020-04-27 09:05:37 UTC  >LOG:  checkpoint starting: shutdown immediate
< 2020-04-27 09:06:04.094 UTC    13970 5ea6a061.3692 2020-04-27 09:05:37 UTC  >LOG:  checkpoint complete: wrote 0 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.000 s, sync=0.000 s, total=0.010 s; sync files=0, longest=0.000 s, average=0.000 s; distance=0 kB, estimate=0 kB
< 2020-04-27 09:06:04.103 UTC    13967 5ea6a061.368f 2020-04-27 09:05:37 UTC  >LOG:  database system is shut down

**Comment by dwsteele** [comment]
There's not a lot to go on here. Do you have a specific statement that causes the issue?

**Comment by mponizil** [comment]
No idea if this is related, but we have seen a few unexpected failures. Some look like [`cache lookup failed for type`](https://github.com/pgaudit/pgaudit/issues/77) and a couple are like this with `Segmentation fault`:

```2020-05-13 14:15:28 UTC:10.0.4.36(59820):www:[3299]:LOG:  AUDIT: SESSION,3,1,WRITE,UPDATE,,,"update ""patients"" set ""stripe_id""='xxx',""stripe_invoice_prefix""='yyy',""updated_at""='2020-05-13T14:15:28.000Z' WHERE id = 'zzz'",<none>
2020-05-13 14:15:28 UTC::@:[22326]:LOG:  server process (PID 3299) was terminated by signal 11: Segmentation fault
2020-05-13 14:15:28 UTC::@:[22326]:DETAIL:  Failed process was running: update "patients" set "stripe_id"='xxx',"stripe_invoice_prefix"='yyy',"updated_at"='2020-05-13T14:15:28.000Z' WHERE id = 'zzz'
2020-05-13 14:15:28 UTC::@:[22326]:LOG:  terminating any other active server processes
2020-05-13 14:15:28 UTC:10.0.4.233(41864):www:[3367]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.4.233(41864):www:[3367]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.4.233(41864):www:[3367]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.4.233(41856):www:[3366]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.4.233(41856):www:[3366]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.4.233(41856):www:[3366]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.3.33(49040):www:[3361]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.3.33(49040):www:[3361]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.3.33(49040):www:[3361]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.3.252(42518):www:[3313]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.3.252(42518):www:[3313]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.3.252(42518):www:[3313]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.3.206(48000):www:[3314]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.3.206(48000):www:[3314]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.3.206(48000):www:[3314]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.4.107(42824):www:[3309]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.4.107(42824):www:[3309]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.4.107(42824):www:[3309]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:20.0.2.137(43048):www:[7412]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:20.0.2.137(43048):www:[7412]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:20.0.2.137(43048):www:[7412]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.3.157(44554):www:[3289]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.3.157(44554):www:[3289]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.3.157(44554):www:[3289]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.3.238(53322):www:[3261]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.3.238(53322):www:[3261]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.3.238(53322):www:[3261]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.4.161(55052):www:[3260]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.4.161(55052):www:[3260]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.4.161(55052):www:[3260]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.3.26(59322):www:[3265]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.3.26(59322):www:[3265]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.3.26(59322):www:[3265]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC:10.0.4.234(45066):www:[3368]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC:10.0.4.234(45066):www:[3368]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC:10.0.4.234(45066):www:[3368]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC::@:[22326]:LOG:  archiver process (PID 22333) exited with exit code 1
2020-05-13 14:15:28 UTC::@:[22332]:WARNING:  terminating connection because of crash of another server process
2020-05-13 14:15:28 UTC::@:[22332]:DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
2020-05-13 14:15:28 UTC::@:[22332]:HINT:  In a moment you should be able to reconnect to the database and repeat your command.
2020-05-13 14:15:28 UTC::@:[22326]:LOG:  all server processes terminated; reinitializing
2020-05-13 14:15:28 UTC::@:[3369]:LOG:  database system was interrupted; last known up at 2020-05-13 14:13:45 UTC
2020-05-13 14:15:28 UTC::@:[3369]:LOG:  database system was not properly shut down; automatic recovery in progress
2020-05-13 14:15:28 UTC::@:[3369]:LOG:  redo starts at 296/F8000828
2020-05-13 14:15:28 UTC::@:[3369]:LOG:  invalid record length at 296/FC00E028: wanted 24, got 0
2020-05-13 14:15:28 UTC::@:[3369]:LOG:  redo done at 296/FC00DFE8
2020-05-13 14:15:28 UTC::@:[3369]:LOG:  last completed transaction was at log time 2020-05-13 14:15:28.265398+00
2020-05-13 14:15:28 UTC::@:[3369]:LOG:  checkpoint starting: end-of-recovery immediate
2020-05-13 14:15:28 UTC::@:[3369]:LOG:  checkpoint complete: wrote 12 buffers (0.0%); 0 WAL file(s) added, 0 removed, 1 recycled; write=0.002 s, sync=0.005 s, total=0.033 s; sync files=12, longest=0.005 s, average=0.000 s; distance=65590 kB, estimate=65590 kB
2020-05-13 14:15:28 UTC::@:[22326]:LOG:  database system is ready to accept connections```

The only obvious commonality at this point is that they've occurred when attempting to insert or update a table that has after insert/update triggers. All the triggers do is insert a row into another table with a snapshot of the change.

**Comment by pyranja** [comment]
Hey! I see a similar behaviour on AWS Aurora Postgresql 11.6:

```
2020-08-29 10:48:35 UTC::@:[28343]:LOG:  server process (PID 28107) was terminated by signal 11: Segmentation fault
2020-08-29 10:48:35 UTC::@:[28343]:DETAIL:  Failed process was running: update "exchange"."balance_locks" set "locked" = ("exchange"."balance_locks"."locked" - ***) where ("exchange"."balance_locks"."account_id" = '*** and "exchange"."balance_locks"."currency_code" = '***' and "exchange"."balance_locks"."order_id" = '***')
```

The table `balance_locks` has an update trigger, that can deletes updated rows if some condition is met.

Additionally the application receives a spurious SQL errors that fail within the update trigger with at least 2 different errors:

```
org.postgresql.util.PSQLException: ERROR: cannot output a value of type pg_ddl_command
Where: SQL statement "DELETE
FROM exchange.balance_locks
WHERE account_id = NEW.account_id
AND order_id = NEW.order_id
AND currency_code = NEW.currency_code"
PL/pgSQL function clear_completed_balance_lock() 
```

```
ERROR: cannot display a value of type index_am_handler
Where: SQL statement "DELETE
FROM exchange.balance_locks
WHERE account_id = NEW.account_id
AND order_id = NEW.order_id
AND currency_code = NEW.currency_code"
```

**Comment by pyranja** [comment]
Saw another variant now in the same setup.

```
Caused by: org.postgresql.util.PSQLException: ERROR: type bigint is not composite
Where: SQL statement "DELETE
FROM exchange.balance_locks
WHERE account_id = NEW.account_id
AND order_id = NEW.order_id
AND currency_code = NEW.currency_code"
PL/pgSQL function clear_completed_balance_lock() 
```

Having segmentation faults and varying error messages looks to me like racey access to memory.
It's also the case, that errors occur spuriously under load.

**Comment by mponizil** [comment]
@pyranja in our situation we are stuck on pgaudit 1.3.0 because [RDS hasn't added support](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_PostgreSQL.html#PostgreSQL.Concepts.General.FeatureSupport.Extensions) for 1.3.1 which seems to have a patch for this issue. For the time being we've disabled the log_parameter flag.

**Comment by pyranja** [comment]
thanks for the heads up @mponizil ! RDS Aurora actually seems to support version 1.3.1, unfortunately we are blocked on the update due to other engine bugs currently >:) 
Good point about the log_parameter - currently trying to work around by suppressing logs on the table where the trigger is used, which seems to work too.

**Comment by dwsteele** [comment]
As already noted, this issue should be fixed in 1.3.1 by 74605a5bc1. Unfortunately, there's not a lot we can do about getting AWS to update.

**Comment by mbanck** [comment]
>  Database works fine when pgaudit is disabled.

With "disabled", do you mean the `pgaudit` extension is dropped or  (rather, I assume) the pgaudit library removed from `shared_preload_libaries` and the instance restarted?
